design/pipeline-artifacts.md

```diff
@@ -37,23 +39,80 @@ Artifacts are distinct from other `resources` types, including `containers`, `re
+
+Example:
+```yaml
+resources:
+  pipelines:
+  - pipeline: mypipe
+
+jobs:
+- job: makeartifact
+  steps:
+  - script: ./build.sh
+  - upload: outputs/**/*
+
+- job: use1artifact
+  dependsOn: makeartifact
+  steps:
+  # by default, a download step is injected -- see later section
+  - script: ls $(Pipeline.Workspace)
+    # this listing shows a folder called `makeartifact`, since the prior job didn't specify a name
+
+- job: use2artifacts
+  dependsOn: makeartifact
+  steps:
+  - download: mypipe   # downloads all artifacts from `mypipe`
+  - download: current  # must include this, since by including a download step, we don't get automatic behavior anymore
+  - script: ls $(Pipeline.Workspace)
+    # this listing shows two folders: `makeartifact` and `mypipe`
+```
+
+#### Controlling download path
+
+If the `path` key is specified, it's a relative path from `$(Pipeline.Workspace)`. Directory names are not automatically injected by the pipeline anymore (but of course, directories present in the artifact itself are still used).
+
+Example:
+```yaml
+jobs:
+- job: makeartifact
+  steps:
+  - script: ./build.sh
+  - upload: outputs/**/*
+
+- job: useartifact
+  dependsOn: makeartifact
+  steps:
+  - download: current
+    path: foo
+  - script: ls $(Pipeline.Workspace)
+    # listing shows one folder, "foo"
+```
+
+#### Build and RM classic pipelines
+No change to current behavior. Artifacts are downloaded to `$(System.DefaultWorkingDirectory)`, which is the sources folder on Build and the artifacts folder on RM.
+
+**Note**: this means we need a new job message type "Pipeline" in addition to the existing "Build" and "Release".
 
 ### Schema
 
 ```yaml
-- download: string # identifier for the pipeline resource from which to download artifacts, optional; "current" means the current pipeline
-  artifact: string # identifier for the artifact to download; required
+- download: string # identifier for the pipeline resource from which to download artifacts, optional; "current" means the current pipeline, blank means all available pipelines (including current)
+  artifact: string # identifier for the artifact to download; optional
   patterns: string # a minimatch path or list of [minimatch paths](tasks/file-matching-patterns.md) to download; if blank, the entire artifact is downloaded
-  root: string # the directory in which to download files, defaults to $PIPELINE_RESOURCESDIRECTORY; if a relative path is provided, it will be rooted from $SYSTEM_DEFAULTWORKINGDIRECTORY
+  path: string # the directory in which to download files, defaults to $(Pipeline.Worksapce); if a relative path is provided, it will be rooted from $(Pipeline.Workspace)
```

Comment:
I think we should start off by always requiring this to be a relative path.

Reply:
:+1: